File: weaving_patterns/WeavingPatterns.java
package com.research.higherweakorder;

import java.util.function.Consumer;

public class WeavingPattern {
    public static void iterate(int n, Consumer<int[][]> onPattern) {
        int[][] P = new int[n][];
        for (int i = 0; i < n; i++) {
            P[i] = new int[n-1];
            for (int j = 0; j < n; j++) {
                if (j < i) {
                    P[i][j] = j;
                } else if (j > i) {
                    P[i][j-1] = j;
                }
            }
        }
        int[][] braids = new int[n][];
        int[][] unbraids = new int[n][];
        for (int i = 0; i < n; i++) {
            braids[i] = new int[n-2];
            unbraids[i] = new int[n-2];
            for (int j = 0; j < n-2; j++) {
                braids[i][j] = 0;
                unbraids[i][j] = 0;
            }
            if (i >= 2) {
                braids[i][i-2] = ((i-2) << 12) | ((i-2) << 8) | ((i-1) << 4) | (i-2);
            }
        }
        _iterate(P, braids, unbraids, onPattern);
    }

    public static int[][] rotate(int[][] P) {
        int n = P.length;
        int[][] ret = new int[n][n-1];
        for (int i = 0; i < n-1; i++) {
            for (int j = 0; j < n-1; j++) {
                ret[i][j] = (P[i+1][j]-1+n) % n;
            }
        }
        for (int i = 0; i < n-1; i++) {
            ret[n-1][i] = P[0][(n-2)-i] - 1;
        }
        return ret;
    }

    public static int[][] reflect(int[][] P) {
        int n = P.length;
        int[][] ret = new int[n][n-1];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n-1; j++) {
                ret[i][j] = (n-1) - P[(n-1)-i][j];
            }
        }

        return ret;
    }

    // P = weaving pattern
    // braids are encoded as (x << 12) | (ix << 8) | (y << 4) | (iy << 4)
    public static void _iterate(int[][] P, int[][] braids, int[][] unbraids, Consumer<int[][]> onPattern) {
        onPattern.accept(P);
        
        int n = P.length;

        for (int z = n-1; z >= 0; z--) {
            for (int iz = 0; iz < n-2; iz++) {
                int braid = braids[z][iz];
                if (braid == 0) {
                    continue;
                }

                int x = (braid >> 12) & 0xf;
                int ix = (braid >> 8) & 0xf;
                int y = (braid >> 4) & 0xf;
                int iy = (braid >> 0) & 0xf;

                // modify P in-place
                P[x][ix] ^= P[x][ix+1];
                P[x][ix+1] ^= P[x][ix];
                P[x][ix] ^= P[x][ix+1];

                P[y][iy] ^= P[y][iy+1];
                P[y][iy+1] ^= P[y][iy];
                P[y][iy] ^= P[y][iy+1];

                P[z][iz] ^= P[z][iz+1];
                P[z][iz+1] ^= P[z][iz];
                P[z][iz] ^= P[z][iz+1];

                // compute new braids
                braids[z][iz] = 0;
                int[] toReset = {-1, -1, -1, -1, -1};
                // xaz
                if (iz+2 < n-1 && 0 < ix && P[x][ix-1] == P[z][iz+2] && P[x][ix-1] < z && P[x][ix-1] > x) {
                    int a = P[x][ix-1];
                    int ia = 0;
                    while (P[a][ia] != x) {
                        ia++;
                    }
                    braids[z][iz+1] = (x << 12) | ((ix-1) << 8) | (a << 4) | (ia << 0);
                    toReset[0] = (z << 4) | (iz+1);
                }
                // ayz
                if (0 < iz && 0 < iy && P[z][iz-1] == P[y][iy-1] && P[y][iy-1] < y) {
                    int a = P[z][iz-1];
                    int ia = 0;
                    while (P[a][ia] != y) {
                        ia++;
                    }
                    braids[z][iz-1] = (a << 12) | (ia << 8) | (y << 4) | ((iy-1) << 0);
                    toReset[1] = (z << 4) | (iz-1);
                }
                // xya
                if (ix+2 < n-1 && iy+1 < n-2 && P[x][ix+2] == P[y][iy+2] && P[x][ix+2] > y) {
                    int a = P[x][ix+2];
                    int ia = 0;
                    while (P[a][ia] != x) {
                        ia++;
                    }
                    braids[a][ia] = (x << 12) | ((ix+1) << 8) | (y << 4) | ((iy+1) << 0);
                    toReset[2] = (a << 4) | (ia);
                }
                // set axy = 0
                if (0 < iy) {
                    int a = P[y][iy-1];
                    toReset[3] = (braids[y][iy-1] << 8) | (y << 4) | (iy-1);
                    braids[y][iy-1] = 0;

                }
                // set yza = 0
                if (iy+2 < n-1) {
                    int a = P[y][iy+2];
                    int ia = 0;
                    while (P[a][ia] != y) {
                        ia++;
                    }
                    if (ia < n-2) {
                        toReset[4] = (braids[a][ia] << 8) | (a << 4) | (ia);
                        braids[a][ia] = 0;
                    }
                }

                // new unbraids
                int[] unbraidsToReset = {-1, -1, -1, -1, -1};
                unbraids[z][iz] = 1;
                // unbraid xya = 0
                if (0 < ix && 0 < iy && P[x][ix-1] == P[y][iy-1] && P[x][ix-1] > y) {
                    int a = P[x][ix-1];
                    int ia = 0;
                    while (P[a][ia] != y) {
                        ia++;
                    }
                    if (ia < n-2) {
                        unbraidsToReset[0] = (unbraids[a][ia] << 8) | (a << 4) | (ia);
                        unbraids[a][ia] = 0;
                    }
                }
                // unbraid xaz = 0
                if (0 < iz && ix+2 < n-1 && P[z][iz-1] == P[x][ix+2] && P[z][iz-1] > x) {
                    unbraidsToReset[1] = (unbraids[z][iz-1] << 8) | (z << 4) | (iz-1);
                    unbraids[z][iz-1] = 0;
                }
                // unbraid ayz = 0
                if (iz+2 < n-1 && iy+2 < n-1 && P[z][iz+2] == P[y][iy+2] && P[z][iz+2] < y) {
                    unbraidsToReset[2] = (unbraids[z][iz+1] << 8) | (z << 4) | (iz+1);
                    unbraids[z][iz+1] = 0;
                }
                // unbraid axy = 1
                if (ix+2 < n-1 && iy+2 < n-1 && P[x][ix+2] == P[y][iy+2] && P[x][ix+2] < x) {
                    unbraidsToReset[3] = (unbraids[y][iy+1] << 8) | (y << 4) | (iy+1);
                    unbraids[y][iy+1] = 1;
                }
                // unbraid yza = 1
                if (0 < iy && 0 < iz && P[y][iy-1] == P[z][iz-1] && P[y][iy-1] > z) {
                    int a = P[y][iy-1];
                    int ia = 0;
                    while (P[a][ia] != z) {
                        ia++;
                    }
                    unbraidsToReset[4] = (unbraids[a][ia] << 8) | (a << 4) | (ia);
                    unbraids[a][ia] = 1;
                }


                boolean valid = true;
                boolean done = false;
                for (int w = n-1; w >= 0 && valid && !done; w--) {
                    for (int iw = 0; iw < n-2 && valid && !done; iw++) {
                        if (unbraids[w][iw] != 0) {
                            if (w == z && iw == iz) {
                                done = true;
                            } else {
                                valid = false;
                            }
                        }
                    }
                }
                if (valid) {
                    _iterate(P, braids, unbraids, onPattern);
                }

                for (int i = 0; i < 5; i++) {
                    if (toReset[i] != -1) {
                        braids[(toReset[i]>>4)&0xf][toReset[i] & 0xf] = toReset[i] >> 8;
                    }
                    if (unbraidsToReset[i] != -1) {
                        unbraids[(unbraidsToReset[i]>>4)&0xf][unbraidsToReset[i] & 0xf] = unbraidsToReset[i] >> 8;
                    }
                }
                unbraids[z][iz] = 0;
                braids[z][iz] = braid;
                // modify P in-place
                P[x][ix] ^= P[x][ix+1];
                P[x][ix+1] ^= P[x][ix];
                P[x][ix] ^= P[x][ix+1];

                P[y][iy] ^= P[y][iy+1];
                P[y][iy+1] ^= P[y][iy];
                P[y][iy] ^= P[y][iy+1];

                P[z][iz] ^= P[z][iz+1];
                P[z][iz+1] ^= P[z][iz];
                P[z][iz] ^= P[z][iz+1];
            }
        }
    }

    public static String toString(int[][] P) {
        String ret = "";
        int n = P.length;
        for (int i = n-1; i >= 0; i--) {
            ret += (i+1) + "|";
            for (int j = 0; j < n-1; j++) {
                ret += (P[i][j]+1) + ",";
            }
            if (i != 0) {
                ret += "\n";
            }
        }
        return ret;
    }
}

File: weaving_patterns/weaving_patterns.py
class WeavingPattern:
    @staticmethod
    def iterate(n, on_pattern):
        """
        Iterate through weaving patterns.
        
        Args:
            n: Size of the pattern
            on_pattern: Callback function to be called for each valid pattern
        """
        # Initialize pattern P
        P = []
        for i in range(n):
            P.append([0] * (n-1))
            for j in range(n):
                if j < i:
                    P[i][j] = j
                elif j > i:
                    P[i][j-1] = j
        
        # Initialize braids and unbraids
        braids = []
        unbraids = []
        for i in range(n):
            braids.append([0] * (n-2))
            unbraids.append([0] * (n-2))
            if i >= 2:
                braids[i][i-2] = ((i-2) << 12) | ((i-2) << 8) | ((i-1) << 4) | (i-2)
        
        WeavingPattern._iterate(P, braids, unbraids, on_pattern)
    
    @staticmethod
    def rotate(P):
        """
        Rotate the weaving pattern.
        
        Args:
            P: Weaving pattern
            
        Returns:
            Rotated pattern
        """
        n = len(P)
        ret = [[0 for _ in range(n-1)] for _ in range(n)]
        
        for i in range(n-1):
            for j in range(n-1):
                ret[i][j] = (P[i+1][j]-1+n) % n
        
        for i in range(n-1):
            ret[n-1][i] = P[0][(n-2)-i] - 1
        
        return ret
    
    @staticmethod
    def reflect(P):
        """
        Reflect the weaving pattern.
        
        Args:
            P: Weaving pattern
            
        Returns:
            Reflected pattern
        """
        n = len(P)
        ret = [[0 for _ in range(n-1)] for _ in range(n)]
        
        for i in range(n):
            for j in range(n-1):
                ret[i][j] = (n-1) - P[(n-1)-i][j]
        
        return ret
    
    @staticmethod
    def _iterate(P, braids, unbraids, on_pattern):
        """
        Internal recursion function for iterating through patterns.
        
        Args:
            P: Current weaving pattern
            braids: Braid encodings
            unbraids: Unbraid encodings
            on_pattern: Callback function for each valid pattern
        """
        on_pattern(P)
        
        n = len(P)
        
        for z in range(n-1, -1, -1):
            for iz in range(n-2):
                braid = braids[z][iz]
                if braid == 0:
                    continue
                
                x = (braid >> 12) & 0xf
                ix = (braid >> 8) & 0xf
                y = (braid >> 4) & 0xf
                iy = (braid >> 0) & 0xf
                
                # Modify P in-place - swap adjacent elements (XOR swap)
                P[x][ix], P[x][ix+1] = P[x][ix+1], P[x][ix]
                P[y][iy], P[y][iy+1] = P[y][iy+1], P[y][iy]
                P[z][iz], P[z][iz+1] = P[z][iz+1], P[z][iz]
                
                # Compute new braids
                braids[z][iz] = 0
                to_reset = [-1, -1, -1, -1, -1]
                
                # xaz
                if iz+2 < n-1 and 0 < ix and P[x][ix-1] == P[z][iz+2] and P[x][ix-1] < z and P[x][ix-1] > x:
                    a = P[x][ix-1]
                    ia = 0
                    while P[a][ia] != x:
                        ia += 1
                    braids[z][iz+1] = (x << 12) | ((ix-1) << 8) | (a << 4) | (ia << 0)
                    to_reset[0] = (z << 4) | (iz+1)
                
                # ayz
                if 0 < iz and 0 < iy and P[z][iz-1] == P[y][iy-1] and P[y][iy-1] < y:
                    a = P[z][iz-1]
                    ia = 0
                    while P[a][ia] != y:
                        ia += 1
                    braids[z][iz-1] = (a << 12) | (ia << 8) | (y << 4) | ((iy-1) << 0)
                    to_reset[1] = (z << 4) | (iz-1)
                
                # xya
                if ix+2 < n-1 and iy+1 < n-2 and P[x][ix+2] == P[y][iy+2] and P[x][ix+2] > y:
                    a = P[x][ix+2]
                    ia = 0
                    while P[a][ia] != x:
                        ia += 1
                    braids[a][ia] = (x << 12) | ((ix+1) << 8) | (y << 4) | ((iy+1) << 0)
                    to_reset[2] = (a << 4) | (ia)
                
                # set axy = 0
                if 0 < iy:
                    a = P[y][iy-1]
                    to_reset[3] = (braids[y][iy-1] << 8) | (y << 4) | (iy-1)
                    braids[y][iy-1] = 0
                
                # set yza = 0
                if iy+2 < n-1:
                    a = P[y][iy+2]
                    ia = 0
                    while P[a][ia] != y:
                        ia += 1
                    if ia < n-2:
                        to_reset[4] = (braids[a][ia] << 8) | (a << 4) | (ia)
                        braids[a][ia] = 0
                
                # New unbraids
                unbraids_to_reset = [-1, -1, -1, -1, -1]
                unbraids[z][iz] = 1
                
                # unbraid xya = 0
                if 0 < ix and 0 < iy and P[x][ix-1] == P[y][iy-1] and P[x][ix-1] > y:
                    a = P[x][ix-1]
                    ia = 0
                    while P[a][ia] != y:
                        ia += 1
                    if ia < n-2:
                        unbraids_to_reset[0] = (unbraids[a][ia] << 8) | (a << 4) | (ia)
                        unbraids[a][ia] = 0
                
                # unbraid xaz = 0
                if 0 < iz and ix+2 < n-1 and P[z][iz-1] == P[x][ix+2] and P[z][iz-1] > x:
                    unbraids_to_reset[1] = (unbraids[z][iz-1] << 8) | (z << 4) | (iz-1)
                    unbraids[z][iz-1] = 0
                
                # unbraid ayz = 0
                if iz+2 < n-1 and iy+2 < n-1 and P[z][iz+2] == P[y][iy+2] and P[z][iz+2] < y:
                    unbraids_to_reset[2] = (unbraids[z][iz+1] << 8) | (z << 4) | (iz+1)
                    unbraids[z][iz+1] = 0
                
                # unbraid axy = 1
                if ix+2 < n-1 and iy+2 < n-1 and P[x][ix+2] == P[y][iy+2] and P[x][ix+2] < x:
                    unbraids_to_reset[3] = (unbraids[y][iy+1] << 8) | (y << 4) | (iy+1)
                    unbraids[y][iy+1] = 1
                
                # unbraid yza = 1
                if 0 < iy and 0 < iz and P[y][iy-1] == P[z][iz-1] and P[y][iy-1] > z:
                    a = P[y][iy-1]
                    ia = 0
                    while P[a][ia] != z:
                        ia += 1
                    unbraids_to_reset[4] = (unbraids[a][ia] << 8) | (a << 4) | (ia)
                    unbraids[a][ia] = 1
                
                # Check validity
                valid = True
                done = False
                for w in range(n-1, -1, -1):
                    for iw in range(n-2):
                        if unbraids[w][iw] != 0:
                            if w == z and iw == iz:
                                done = True
                            else:
                                valid = False
                        if not valid or done:
                            break
                    if not valid or done:
                        break
                
                if valid:
                    WeavingPattern._iterate(P, braids, unbraids, on_pattern)
                
                # Reset
                for i in range(5):
                    if to_reset[i] != -1:
                        braids[(to_reset[i] >> 4) & 0xf][to_reset[i] & 0xf] = to_reset[i] >> 8
                    if unbraids_to_reset[i] != -1:
                        unbraids[(unbraids_to_reset[i] >> 4) & 0xf][unbraids_to_reset[i] & 0xf] = unbraids_to_reset[i] >> 8
                
                unbraids[z][iz] = 0
                braids[z][iz] = braid
                
                # Restore P - swap back
                P[x][ix], P[x][ix+1] = P[x][ix+1], P[x][ix]
                P[y][iy], P[y][iy+1] = P[y][iy+1], P[y][iy]
                P[z][iz], P[z][iz+1] = P[z][iz+1], P[z][iz]
    
    @staticmethod
    def to_string(P):
        """
        Convert pattern to string representation.
        
        Args:
            P: Weaving pattern
            
        Returns:
            String representation of pattern
        """
        ret = ""
        n = len(P)
        for i in range(n-1, -1, -1):
            ret += str(i+1) + "|"
            for j in range(n-1):
                ret += str(P[i][j]+1) + ","
            if i != 0:
                ret += "\n"
        return ret

N = 7
with open(f"weaving_patterns_{N}.txt", "w") as f:
    WeavingPattern.iterate(N, lambda P: f.writelines(WeavingPattern.to_string(P)+"\n\n"))